### ISSUE TITLE
495 symlinks

### ISSUE THREAD

**Issue by FakeShemp** [flathub/org.freedesktop.Platform.GL.nvidia#76]
The new 495 beta driver seems to need some files symlinked to work properly.

Details are here: https://www.nvidia.com/download/driverResults.aspx/181167/en

Currently this happens:

```
MESA-LOADER: failed to open nvidia-drm: /usr/lib/x86_64-linux-gnu/GL/default/lib/gbm/nvidia-drm_gbm.so: cannot open shared object file: No such file or directory (search paths /usr/lib/x86_64-linux-gnu/GL/default/lib/gbm, suffix _gbm)
```

**Comment by FakeShemp** [comment]
Looking at RPM Fusion, there also seem to be a `libnvidia-vulkan-producer.so` lib that might need a symlink to help Vulkan.

https://pkgs.rpmfusion.org/cgit/nonfree/xorg-x11-drv-nvidia.git/commit/?id=8b88d958f28cb28f0fcfef8581adfae565212135

**Comment by gasinvein** [comment]
As I wrote in the freedesktop-sdk issue comment, I've managed to get the current GL extension to work with Fedora flatpak runtime. So the issue must be with something in freedesktop-sdk; I'm not sure if the GL extension needs anything else on top of what was already added.

**Comment by gasinvein** [comment]
@FakeShemp I've did some more tests and it appears thet you're right, the `libnvidia-vulkan-producer.so` symlink is required indeed.

**Comment by Gert-dev** [comment]
> So, GPU acceleration doesn't work for native wayland apps in flatpak only, correct?

Is this also the reason that with Wayland on NVIDIA and GBM, the Flatpaked Firefox 96 (using the native Wayland back end, which it uses by default for me) from Flathub falls back to WebRender (Software) and only shows llvmpipe as GPU?

(It works fine when using the non-Flatpak-ed Firefox, as well as on X11 with either the Flatpak-ed or non-Flatpak-ed version.)

**Comment by gasinvein** [comment]
It should be fixed in the freedesktop runtime 21.08.9 (released on 10 Jan or so).

**Comment by prodrigestivill** [comment]
I think that freedesktop runtime 21.08.9 and above has it working using EGL+GBM.

But it fails whem using directly libgbm.

As an example it can be done using Steam+gamescope:
```
flatpak install com.valvesoftware.Steam com.valvesoftware.Steam.Utility.gamescope
```

Then an example of failing GBM:
```sh
flatpak run --command=gamescope com.valvesoftware.Steam -e -- steam -gamepadui
```
Outputs the follwing error:
```
...
couldn't create gbm device
EGL setup failed, disabling glamor
Failed to initialize glamor, falling back to sw
...
```


But setting the `GBM_BACKENDS_PATH` works:
```
flatpak run --env="GBM_BACKENDS_PATH=/usr/lib/x86_64-linux-gnu/GL/$(flatpak --gl-drivers | head -1)/extra/gbm/" --command=gamescope com.valvesoftware.Steam -e -- steam -gamepadui
```

**Comment by zdobersek** [comment]
With the `nvidia-525-89-02` version, the `nvidia-drm_gbm.so` symlink is set up under `/usr/lib/x86_64-linux-gnu/GL/nvidia-525-89-02/extra/gbm/`.  I too can confirm exporting `GBM_BACKENDS_PATH` brings things together.

But originally the default gbm backends path is used. That is a configurable option on the libgbm project. This is where the `nvidia-drm_gbm.so` binary is looked for by default:
```
strace.log.482:03:13:07.306372 openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/gbm/tls/nvidia-drm_gbm.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
strace.log.482:03:13:07.306389 openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/gbm/nvidia-drm_gbm.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

So putting a `nvidia-drm_gbm.so` link there would make things work out of the box.

FWIW the code doing the loading is in Mesa:
https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/loader/loader.c#L620
https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gbm/main/backend.c#L195

**Comment by TingPing** [comment]
I believe there isn't anything this package can do. This needs to be handled in [freedesktop-sdk](https://gitlab.com/freedesktop-sdk/freedesktop-sdk).

They could perhaps patch libgbm or set up symlinks that point into the extensions.

**Comment by doraskayo** [comment]
> I believe there isn't anything this package can do. This needs to be handled in [freedesktop-sdk](https://gitlab.com/freedesktop-sdk/freedesktop-sdk).
> 
> They could perhaps patch libgbm, set up symlinks that point into the extensions, or even set the env var by default.

No need to patch. There's a `gbm-backends-path` Meson option: https://gitlab.freedesktop.org/mesa/mesa/-/blob/be4f7fb656180ab55a50eff01f36672b0bf5f146/meson_options.txt#L314-320

This also requires the `gbm` directory to be added to `merge-dirs` in `freedesktop-sdk`.

If anyone is interested in having this functionality and is willing to test it thoroughly, an issue requesting this should be opened in https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues.

**Comment by doraskayo** [comment]
This should already be working in `23.08beta` ([freedesktop-sdk!14619](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/merge_requests/14619)) and will also be available in the next release of `22.08` ([freedesktop-sdk!14640](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/merge_requests/14640)).

**Comment by guihkx** [comment]
Closing as fixed as per the comment above.
